### ISSUE TITLE
Activate fiat onramp in Torus and check out our options

### ISSUE THREAD

**Issue by jamespfarrell** [Giveth/giveth-2#435]
## 📋 Anything to add?
Torus has fiat onramping built in. 

It would be easy and fun to activate and assess it.

Any takers?
@geleeroyale @aminlatifi @mateodaza

**Comment by mateodaza** [comment]
COOL!!! is this it? https://docs.tor.us/torus-wallet/api-reference/topup

We may need a simple design to let the user choose wyre, moonpay or rampnetwork @markoprljic  what do you think?

I am considering these two ways:
1. Let whatever user (logged or not) to use the top up service and send crypto to the project's wallet directly paying with fiat
2. Let logged in users to top up their torus wallet

I believe if we do the first option it's enough for MVP. Let's discuss! According to the docs is pretty straightforward

**Comment by willyogo** [comment]
The first option sounds awesome for MVP! A couple q's: 
1. What are the benefits of using tor.us' top up API vs. integrating something like Transak.com or Ramp.network directly?  
2. Do we need to ask users to choose between Wyre, Moonpay, or Ramp.network? If we must pass one of these providers through to tor.us, ideally we wouldn't confuse users by asking users to select a provider, but rather we would ask them to select a payment method & currency, and then automatically select the best provider. For the initial implementation we could also start by hardcoding a single provider from this list.

**Comment by markoprljic** [comment]
Did you check what we already have in Figma?

**Comment by mateodaza** [comment]
For the first questionn

> The first option sounds awesome for MVP! A couple q's:
> 
> 1. What are the benefits of using tor.us' top up API vs. integrating something like Transak.com or Ramp.network directly?
> 2. Do we need to ask users to choose between Wyre, Moonpay, or Ramp.network? If we must pass one of these providers through to tor.us, ideally we wouldn't confuse users by asking users to select a provider, but rather we would ask them to select a payment method & currency, and then automatically select the best provider. For the initial implementation we could also start by hardcoding a single provider from this list.

1. The ease of use, we'd only call a function to set the provider, destination address and some config on amount and crypto. This is only worth it if we will offer the three options, if we choose only Ramp network for example, probably is better to go with their api directly and avoid this.

2. For this top up, it isn't required to be logged in with torus, users can buy crypto for any address, in the first scenario I mention they would be technically buying crypto for the project's wallet. 
If we go with only one provider, donors would go directly to the provider's flow without having to choose.

**Comment by jamespfarrell** [comment]
I don't like the idea of them paying them directly. As we will lose oversight of the transaction in our dapp, I prefer we load their torus wallet, or even any other kind of wallet, then do payment through the dapp.

I would feel differently if were generating a wallet for them, like a Gnosis safe or any kind of Giveth wallet....

**Comment by geleeroyale** [comment]
Lots of discussion - still needs more clarification.

**Comment by jamespfarrell** [comment]
updates from asking on telegram:

Ramp.network is working on offramping right now, so this isn't ready for us.

We need to do research on Simplex and Mercuryo - which have been recommended to us in the Torus telegram.

Core questions that need to be answered are include the use of xDAI.

Can we ramp on and off to xDAI?

What are implications of that?

Can we have both xDAI and mainnet ramping?

What happens if payment is sent to an xDAI address that doesn't exist yet? Is it created at the time that the project logs into xDAI?

**Comment by jamespfarrell** [comment]
In addition to integrating some sort of widget, what are the implications of to our dapp? Will we need to save the onramp transaction in our database somewhere? Will we need to receive an onramp successful ping and update a balance? or something else?

**Comment by willyogo** [comment]
> In addition to integrating some sort of widget, what are the implications of to our dapp? Will we need to save the onramp transaction in our database somewhere? Will we need to receive an onramp successful ping and update a balance? or something else?

Good q's. Even after the user has completed the on ramp flow, it can take a while for the funds to arrive on chain.

I think this would be sufficient for MVP:
1. User completes the payment flow with the onramp
2. User sees donation success screen
3. (current functionality) Once the donation is completed on chain, it shows up in a project's donation totals and history

**Comment by geleeroyale** [comment]
My research (WIP) into this lives in our Notion workspace -> https://www.notion.so/giveth/Fiat-on-off-ramp-00196f93f2af48fda27f78bcdbd864a5

**Comment by Gaiaguardian** [comment]
Off ramping possibility, I am connected to them: gsr.io

**Comment by jamespfarrell** [comment]
> User completes the payment flow with the onramp
> User sees donation success screen
> (current functionality) Once the donation is completed on chain, it shows up in a project's donation totals and history

I'm not sure we can onramp and send a transaction on behalf of the user once the onramp is complete?

We might need for them to return and make the donation after the fiat has been ramped on?

Maybe you know about that by now @geleeroyale ?

> My research (WIP) into this lives in our Notion workspace -> https://www.notion.so/giveth/Fiat-on-off-ramp-00196f93f2af48fda27f78bcdbd864a5

Nice, thanks for this @geleeroyale, what's your feeling?
After a quick scan looks like ramp.network remains top for me since we can onboard to xDai

**Comment by geleeroyale** [comment]
I am leaning towards ramp.network as well. We could provide a different offramp until theirs is ready? (like mooni.tech)

**Comment by GriffGreen** [comment]
> What happens if payment is sent to an xDAI address that doesn't exist yet? Is it created at the time that the project logs into xDAI?

If someone has an ethereum key, they have an xDai key... I am assuming Torus doesn't separate keys per network... The only thing to fear is if someone is using a contract address... like Giveth's multisig... that would be fucked up, the money would be lost. In general, for DApps, a lot of issues can arise when people use smart contracts for interactions... Gnosis Safes, Argent Wallets, multisigs and DAOs... it creates so many problems over time :-( 

We can detect if people are using an account... there is a pretty simple way to do that: 
```
web3.eth.getCode("0x839395e20bbB182fa440d08F850E6c7A8f6F0780")
"0x"
```
You need to be pinging a full node to make that happen... but if the answer is 0x, there is no data sitting at that address... There are some edge cases, where it could be a contract... but all of those edge cases basically mean that the person can deploy the contract on xDai too. So it is a fine check for our purposes.

**Comment by jamespfarrell** [comment]
> If someone has an ethereum key, they have an xDai key... I am assuming Torus doesn't separate keys per network... The only thing to fear is if someone is using a contract address... like Giveth's multisig... that would be fucked up, the money would be lost if xDai was sent to that address.

That's what I thought, but does that also mean that if someone send xDai to that address before it's created that it will be there when someone uses the private key for the first time on xDaiChain?

How do we see this going down?

They click on pay with credit card, instead of crypto, and we send them through the ramp.network widget onto xDai?
![image](https://user-images.githubusercontent.com/10499048/108202643-ba656a00-7121-11eb-824d-d02d6ab0f39a.png)

Then what?

I believe most likely, they will then need to sign the transaction after on-ramping, and do the donation right? 

So it'd be a 2 step process....?

I've asked what can do done about this in Discord...

I believe it would be helpful if we reached out ot

**Comment by jamespfarrell** [comment]
![image.png](https://images.zenhubusercontent.com/5807a7a095e42b6c5baabbd3/bfa4a61e-7107-4778-b606-5f4b4076e1d0)

[Above from Ramp.network discord](https://discord.gg/9hZZ848T)

**Comment by geleeroyale** [comment]
From dev call: Integrate ramp.network -> xdai

**Comment by jamespfarrell** [comment]
One critical thing to consider IMHO is that a user may want to onramp $1000 and distribute it to 10 different projects... and I think that's definitely something we should support.

**Comment by jamespfarrell** [comment]
So I believe that the fastest easiest route for us for Fiat onramping is via [Torus topup functionality](https://docs.tor.us/torus-wallet/api-reference/topup)

Though it is a similar set up apparently:
![image](https://user-images.githubusercontent.com/10499048/108275605-df37fc80-7176-11eb-86ba-afa394eacbf4.png)

**Comment by mateodaza** [comment]
The problem with Torus topup is that we only get a payment confirmation, is that enough to consider it a donation? - I believe if we go directly to ramp network we can use their webhooks to get the tx hash

**Comment by jamespfarrell** [comment]
![image](https://user-images.githubusercontent.com/10499048/108276346-f7f4e200-7177-11eb-9c4d-059196552b16.png)

![image](https://user-images.githubusercontent.com/10499048/108276369-00e5b380-7178-11eb-8a2e-300dce4817cc.png)

**Comment by jamespfarrell** [comment]
Exploring Torus 
![image](https://user-images.githubusercontent.com/10499048/108333535-5601e380-71d1-11eb-8235-8b44c36ce679.png)

Click next opens popup:
![image](https://user-images.githubusercontent.com/10499048/108333629-7336b200-71d1-11eb-872e-16c7a6393a22.png)

**Comment by jamespfarrell** [comment]
![image](https://user-images.githubusercontent.com/10499048/108339420-b1cf6b00-71d7-11eb-956e-5286a3acf1f8.png)
Fiat on-ramping appears to already be in place for Torus?

**Comment by jamespfarrell** [comment]
ok we have a compliance issue with ramping directly to the project

**Comment by willyogo** [comment]
Dang.. I was worried about this 😢 

I had reached out to https://transak.com/ with this same question previously and they said it is possible with their services and not a violation of their terms. Perhaps their providers have different licensing than ramp.network (ie. money transmitter license)? Transak does not support xDAI but does support a handful of other networks. I really like them too, can we consider using them? We would need to KYC Giveth, maybe next step is to see if we can get approved. Happy to take that on if we're interested

https://integrate.transak.com/defaultCryptoAmount-668b916b7b95421a8ebd33f4ffb686f3

![image](https://user-images.githubusercontent.com/26232238/108387957-5d25f300-71cb-11eb-9fd2-1bc451dd599d.png)

![image](https://user-images.githubusercontent.com/26232238/108388067-7af35800-71cb-11eb-945d-6a17a70d70f8.png)

![image](https://user-images.githubusercontent.com/26232238/108388094-847cc000-71cb-11eb-81f0-8413cc19c772.png)

**Comment by jamespfarrell** [comment]
So we can only ramp onto the wallet of the logged in user. I'm wondering how best do this. 
If they click on 'pay with card', then we could explain to them that they need to pre-load their account like a prepayed card. Or something. Then we onboard the they fiat into their account, receive a webhook response and then email them to tell them a link to a prepopulated donate with crypto form or something....
@markop @willyfox everyone else?

**Comment by jamespfarrell** [comment]
Email prepopulated for uses logged in with Torus:
![image](https://user-images.githubusercontent.com/10499048/108394080-9c2c6680-7214-11eb-9137-e1fe4aabc97b.png)

![image](https://user-images.githubusercontent.com/10499048/108394173-bbc38f00-7214-11eb-9e6d-82f82da08141.png)

If I click add new :
![image](https://user-images.githubusercontent.com/10499048/108394290-d990f400-7214-11eb-969d-da72e9194a98.png)

![image](https://user-images.githubusercontent.com/10499048/108394457-06450b80-7215-11eb-9d5b-dde31dbd3747.png)

![image](https://user-images.githubusercontent.com/10499048/108394535-16f58180-7215-11eb-9805-c6435ea44d7a.png)

![image](https://user-images.githubusercontent.com/10499048/108394555-1fe65300-7215-11eb-8ea1-0742ca458f76.png)

Transaction summary 👎 
![image](https://user-images.githubusercontent.com/10499048/108394627-32f92300-7215-11eb-8715-8bd1f3097ef1.png)

**Comment by jamespfarrell** [comment]
When they click back to dapp we can explain that they will be notified when that has gone through
We will receive a webhook call, and email them with a direct donate with crypto link

**Comment by jamespfarrell** [comment]
[transaction details](https://ri-dashboard-staging.firebaseapp.com/#/details/sohkaneqv9vcsbn?secret=j5b5krdfgzyuvjc2)

**Comment by willyogo** [comment]
Let me double clarify with transak.com that this is indeed acceptable in their and their partner's terms. I think the topping up is a creative solution and am open to it if it's our best option, but prefer the flow of donating directly to a project if that's possible. I'm assuming the more common use case will be somebody wanting to donate to one specific project (as opposed to preloading, waiting for the funds to arrive, and then completing several donations).

**Comment by willyogo** [comment]
Hopefully will get a response soon 
![image](https://user-images.githubusercontent.com/26232238/108416714-565ba800-71ec-11eb-9085-c5777a0237c5.png)

**Comment by willyfox** [comment]
You must have made a mistake including me as a recipient. My github
username has nothing to do with your team member name :) I am not part of
this project. Please kindly pull me off this thread. Cheers

**Comment by jamespfarrell** [comment]
I have activated default Torus onramping. There is a separate issue for documenting this process.
